### PR TITLE
refactor: delete iterations and their sub items together

### PR DESCRIPTION
**Pull request by jing-zhi**
#### What this PR does / why we need it:
refactor: delete iterations and their sub items together
feat: initialize issueService

#### Which issue(s) this PR fixes:
refactor: delete iterations and their sub items together

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/iteration/12783/all?id=502433&iterationID=12783&type=REQUIREMENT)

#### Specified Reviewers:

/assign @chengjoey @sfwn @luobily 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |      refactor: delete iterations and their sub items together        |
| 🇨🇳 中文    |      【项目协同】删除迭代支持选择确认删除迭代中所有事项并删除迭代        |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: internal/apps/dop/endpoints/iteration.go [hunk 1]
```diff
@@ -235,6 +237,33 @@ func (e *Endpoints) DeleteIteration(ctx context.Context, r *http.Request, vars m
 		return errorresp.ErrResp(err)
 	}
 
+	// 检查 iteration 下是否有需求/任务/bug;
```
Comment (sfwn) [inline internal/apps/dop/endpoints/iteration.go:240]: use English

File: internal/apps/dop/endpoints/iteration.go [hunk 2]
```diff
@@ -235,6 +237,33 @@ func (e *Endpoints) DeleteIteration(ctx context.Context, r *http.Request, vars m
...
+	issueReq := pb.PagingIssueRequest{
+		ProjectID:   iteration.ProjectID,
+		IterationID: int64(iterationID),
+		External:    true,
+		PageNo:      1,
+		PageSize:    10000,
```
Comment (sfwn) [inline internal/apps/dop/endpoints/iteration.go:246]: Is there a parameter to only retrieve the total, as retrieving issue details consumes a lot of resources.

File: internal/apps/dop/endpoints/iteration.go [hunk 3]
```diff
@@ -235,6 +237,33 @@ func (e *Endpoints) DeleteIteration(ctx context.Context, r *http.Request, vars m
...
+	if len(issues) > 0 {
+		itemIDs := []string{}
+		for _, v := range issues {
+			itemIDs = append(itemIDs, strconv.FormatInt(v.Id, 10))
+		}
+		var req = &pb.BatchDeleteIssueRequest{}
+		req.Ids = itemIDs
+		req.ProjectID = issues[0].ProjectID
+		ctx = apis.WithUserIDContext(ctx, identityInfo.UserID)
+		_, err = e.issueService.BatchDeleteIssues(ctx, req)
```
Comment (sfwn) [inline internal/apps/dop/endpoints/iteration.go:261]: Just delete by iteration id.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: internal/apps/dop/endpoints/iteration.go (hunk 1 above)
Comment (jing-zhi) [reply internal/apps/dop/endpoints/iteration.go]: ok,thanks

File: internal/apps/dop/endpoints/iteration.go (hunk 2 above)
Comment (jing-zhi) [reply internal/apps/dop/endpoints/iteration.go:247]: No, but your proposal is great. I plan to write a function that iteratively obtains a list of issue IDs

File: internal/apps/dop/endpoints/iteration.go (hunk 3 above)
Comment (jing-zhi) [reply internal/apps/dop/endpoints/iteration.go]: ok, thanks

File: api/proto/openapiv1/dop/dop.proto [hunk 4]
```diff
@@ -409,6 +409,7 @@ message CMDB_CERTIFICATE_GET_Request {
 }
 message CMDB_ITERATION_DELETE_Request {
     string id = 1; // generated from path variable: id. You should change the proto type if necessary.
+    bool  onlyIteration = 2;
```
Comment (sfwn) [inline api/proto/openapiv1/dop/dop.proto:412]: format it

Comment (sfwn) [reply api/proto/openapiv1/dop/dop.proto:412]: protos under `openapiv1` dir is auto generated, so don't modify it manually.

File: api/proto/openapiv1/dop/dop.proto [hunk 5]
```diff
@@ -2905,10 +2906,10 @@ service dop {
     };
     rpc CMDB_ITERATION_DELETE (CMDB_ITERATION_DELETE_Request) returns (google.protobuf.Empty) {
         option (google.api.http) = {
-            delete: "/api/iterations/{id}",
+            delete: "/api/iterations/{id}/{onlyIteration}",
```
Comment (sfwn) [inline api/proto/openapiv1/dop/dop.proto:2909]: this change will let original proxy invalid.

File: internal/apps/dop/endpoints/endpoints.go [hunk 6]
```diff
@@ -462,7 +463,7 @@ func (e *Endpoints) Routes() []httpserver.Endpoint {
 
 		{Path: "/api/iterations", Method: http.MethodPost, Handler: e.CreateIteration},
 		{Path: "/api/iterations/{id}", Method: http.MethodPut, Handler: e.UpdateIteration},
-		{Path: "/api/iterations/{id}", Method: http.MethodDelete, Handler: e.DeleteIteration},
+		{Path: "/api/iterations/{id}/{onlyIteration}", Method: http.MethodDelete, Handler: e.DeleteIteration},
```
Comment (sfwn) [inline internal/apps/dop/endpoints/endpoints.go:466]: this change will let original proxy invalid.

File: internal/apps/dop/endpoints/iteration.go [hunk 7]
```diff
@@ -235,6 +237,27 @@ func (e *Endpoints) DeleteIteration(ctx context.Context, r *http.Request, vars m
...
 	}
 
+	itemIDs, err := e.issueService.DBClient().GetIssuesIDByIterationID(iterationID)
+	if err != nil {
+		return apierrors.ErrDeleteIteration.InternalError(err).ToResp(), nil
+	}
+
+	if len(itemIDs) > 0 {
+		if onlyIteration {
+			err = e.issueService.DBClient().BatchUpdateIssueIterationID(itemIDs, -1)
```
Comment (sfwn) [inline internal/apps/dop/endpoints/iteration.go:247]: args should support `original iterationID`.

File: internal/apps/dop/providers/issue/core/service.go [hunk 8]
```diff
@@ -616,25 +612,30 @@ func (i *IssueService) BatchDeleteIssues(ctx context.Context, req *pb.BatchDelet
 		// 删除测试计划用例关联
 		if issue.Type == pb.IssueTypeEnum_BUG {
 			if err := i.db.DeleteIssueTestCaseRelationsByIssueIDs([]uint64{numSlice[k]}); err != nil {
```
Comment (sfwn) [inline internal/apps/dop/providers/issue/core/service.go:614]: I think tx not works here.

Comment (sfwn) [reply internal/apps/dop/providers/issue/core/service.go:614]: you should pass tx as args.

File: internal/apps/dop/providers/issue/dao/core.go [hunk 9]
```diff
@@ -161,6 +161,10 @@ func (client *DBClient) UpdateIssues(requirementID uint64, fields map[string]int
 		Updates(fields).Error
 }
 
+func (client *DBClient) BatchUpdateIssueIterationID(ids []uint64, iterationID int64) error {
+	return client.Model(Issue{}).Where("deleted = 0").Where("id IN (?)", ids).Updates(Issue{IterationID: iterationID}).Error
```
Comment (sfwn) [inline internal/apps/dop/providers/issue/dao/core.go:165]: - you'd better check ids firstly
- use map in `Updates` is better (bool are easily overlooked)

File: internal/apps/dop/providers/issue/dao/core.go [hunk 10]
```diff
@@ -226,6 +230,11 @@ func (client *DBClient) BatchDeleteIssues(ids []uint64) error {
 	return client.Model(Issue{}).Where("id IN (?)", ids).Update("deleted", 1).Error
 }
 
+func (client *DBClient) BatchDeleteIssueByIterationID(iterationID uint64) error {
+	sql := fmt.Sprintf("UPDATE dice_issues AS dis LEFT JOIN dice_issue_testcase_relations AS ditr ON ditr.issue_id = dis.id LEFT JOIN dice_issue_relation AS dir ON dir.issue_id = dis.id LEFT JOIN dice_issue_property_relation AS dipr ON dipr.issue_id = dis.id LEFT JOIN erda_issue_state_transition AS eist ON eist.issue_id = dis.id SET dis.deleted = 1 WHERE dis.iteration_id = ?;")
```
Comment (sfwn) [inline internal/apps/dop/providers/issue/dao/core.go:234]: Format this SQL into multiple lines

File: internal/core/openapi/legacy/api/apis/dop/cmdb_iteration_delete.go [hunk 11]
```diff
@@ -23,8 +23,8 @@ import (
 )
 
 var CMDB_ITERATION_DELETE = apis.ApiSpec{
-	Path:        "/api/iterations/<id>",
-	BackendPath: "/api/iterations/<id>",
+	Path:        "/api/iterations/<id>/<onlyIteration>",
```
Comment (sfwn) [inline internal/core/openapi/legacy/api/apis/dop/cmdb_iteration_delete.go:26]: bad

**Review by sfwn — CHANGES_REQUESTED** [review]

File: internal/apps/dop/providers/issue/dao/core.go (hunk 9 above)
Comment (jing-zhi) [reply internal/apps/dop/providers/issue/dao/core.go:165]: ok,thanks

File: internal/core/openapi/legacy/api/apis/dop/cmdb_iteration_delete.go (hunk 11 above)
Comment (jing-zhi) [reply internal/core/openapi/legacy/api/apis/dop/cmdb_iteration_delete.go:26]: ok，Modified

File: internal/apps/dop/endpoints/endpoints.go (hunk 6 above)
Comment (jing-zhi) [reply internal/apps/dop/endpoints/endpoints.go:466]: ok，modified

File: api/proto/openapiv1/dop/dop.proto (hunk 5 above)
Comment (jing-zhi) [reply api/proto/openapiv1/dop/dop.proto:2909]: ok，modified

File: api/proto/openapiv1/dop/dop.proto (hunk 4 above)
Comment (jing-zhi) [reply api/proto/openapiv1/dop/dop.proto:412]: ok，modified

File: internal/apps/dop/providers/issue/core/service.go (hunk 8 above)
Comment (jing-zhi) [reply internal/apps/dop/providers/issue/core/service.go:614]: ok，i reinitialized the parameters using tx

File: internal/apps/dop/endpoints/iteration.go [hunk 12]
```diff
@@ -205,6 +206,7 @@ func (e *Endpoints) DeleteIteration(ctx context.Context, r *http.Request, vars m
 	}
 
 	iterationID, err := strconv.ParseUint(vars["id"], 10, 64)
+	onlyIteration := r.URL.Query().Get("onlyIteration")
```
Comment (sfwn) [inline internal/apps/dop/endpoints/iteration.go:209]: parse `onlyIteration` string to bool here

File: internal/apps/dop/endpoints/iteration.go [hunk 13]
```diff
@@ -235,6 +237,27 @@ func (e *Endpoints) DeleteIteration(ctx context.Context, r *http.Request, vars m
...
+
+	if len(itemIDs) > 0 {
+		ctx = apis.WithUserIDContext(ctx, identityInfo.UserID)
+		if onlyIteration == "true" {
+			err = e.issueService.BatchUpdateIssueIterationIDByIterationID(ctx, iterationID, -1)
+			if err != nil {
+				return apierrors.ErrDeleteIteration.InternalError(err).ToResp(), nil
+			}
+			goto come
+		} else if onlyIteration == "false" {
```
Comment (sfwn) [inline internal/apps/dop/endpoints/iteration.go:253]: no need else if, all non-true treat as `false`.

File: internal/apps/dop/providers/issue/core/service.go [hunk 14]
```diff
@@ -646,10 +650,40 @@ func (i *IssueService) BatchDeleteIssues(ctx context.Context, req *pb.BatchDelet
...
 	return &pb.BatchDeleteIssueResponse{Data: issues}, nil
 }
 
+// BatchDeleteIssueByIterationID 根据迭代 id 批量删除 issue
+func (i *IssueService) BatchDeleteIssueByIterationID(ctx context.Context, iterationID uint64) (err error) {
+	identityInfo := apis.GetIdentityInfo(ctx)
+	if identityInfo == nil {
+		return apierrors.ErrUpdateIssue.NotLogin()
+	}
+	tx := i.db.Begin()
```
Comment (sfwn) [inline internal/apps/dop/providers/issue/core/service.go:663]: no need tx here, only one operation.

File: internal/apps/dop/providers/issue/core/service.go [hunk 15]
```diff
@@ -646,10 +650,40 @@ func (i *IssueService) BatchDeleteIssues(ctx context.Context, req *pb.BatchDelet
...
+	return
+}
+
+// BatchUpdateIssueIterationIDByIterationID 根据迭代 id 批量更新 issue 的 iteration_id
+func (i *IssueService) BatchUpdateIssueIterationIDByIterationID(ctx context.Context, iterationID uint64, ID int64) (err error) {
+	identityInfo := apis.GetIdentityInfo(ctx)
+	if identityInfo == nil {
+		return apierrors.ErrUpdateIssue.NotLogin()
+	}
+	tx := i.db.Begin()
```
Comment (sfwn) [inline internal/apps/dop/providers/issue/core/service.go:678]: same here.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: internal/apps/dop/endpoints/iteration.go (hunk 13 above)
Comment (sfwn) [reply internal/apps/dop/endpoints/iteration.go:253]: and no need `goto come`

File: internal/apps/dop/endpoints/iteration.go (hunk 12 above)
Comment (jing-zhi) [reply internal/apps/dop/endpoints/iteration.go:209]: ok,thanks

File: internal/apps/dop/endpoints/iteration.go (hunk 13 above)
Comment (jing-zhi) [reply internal/apps/dop/endpoints/iteration.go:253]: OK，thanks

File: internal/apps/dop/providers/issue/core/service.go (hunk 14 above)
Comment (jing-zhi) [reply internal/apps/dop/providers/issue/core/service.go:663]: Oh, yes, I neglected

File: internal/apps/dop/providers/issue/core/service.go (hunk 15 above)
Comment (jing-zhi) [reply internal/apps/dop/providers/issue/core/service.go:678]: Oh, yes, I neglected

File: internal/apps/dop/endpoints/iteration.go [hunk 16]
```diff
@@ -235,6 +237,27 @@ func (e *Endpoints) DeleteIteration(ctx context.Context, r *http.Request, vars m
...
+	}
+
+	if len(itemIDs) > 0 {
+		ctx = apis.WithUserIDContext(ctx, identityInfo.UserID)
+		if onlyIteration {
+			err = e.issueService.BatchUpdateIssueIterationIDByIterationID(ctx, iterationID, -1)
+			if err != nil {
+				return apierrors.ErrDeleteIteration.InternalError(err).ToResp(), nil
+			}
+			goto come
```
Comment (sfwn) [inline internal/apps/dop/endpoints/iteration.go:252]: necessary `goto` instruct.

**Review by sfwn — CHANGES_REQUESTED** [review]

**Review by sfwn — APPROVED** [review]

**Pull request merged**